src/libsync/pushnotifications.cpp

```diff
@@ -0,0 +1,112 @@
+#include "pushnotifications.h"
+#include "creds/abstractcredentials.h"
+#include "websocket.h"
+#include "account.h"
+
+namespace OCC {
+
+PushNotifications::PushNotifications(Account *account, QSharedPointer<AbstractWebSocket> webSocket)
+    : _account(account)
+    , _webSocket(webSocket)
+{
+}
+
+PushNotificationTypes PushNotifications::pushNotificationsAvailable() const
+{
+    return _account->capabilities().pushNotificationsAvailable();
+}
+
+void PushNotifications::reconnect()
+{
+    connectWebSocket();
+}
+
+void PushNotifications::connectWebSocket()
+{
+    disconnectWebSocket();
+
+    setupFilesPushNotifications();
+
+    openWebSocket();
+}
+
+void PushNotifications::disconnectWebSocket()
+{
+    _webSocket->close();
+}
+
+void PushNotifications::onWebSocketConnected()
+{
+    disconnect(_webSocket.data(), &AbstractWebSocket::textMessageReceived, this, &PushNotifications::onWebSocketTextMessageReceived);
+    connect(_webSocket.data(), &AbstractWebSocket::textMessageReceived, this, &PushNotifications::onWebSocketTextMessageReceived);
+
+    authenticateOnWebSocket();
+}
+
+void PushNotifications::authenticateOnWebSocket()
+{
+    const auto credentials = _account->credentials();
+    const auto username = credentials->user();
+    const auto password = credentials->password();
+
+    // Authenticate
+    _webSocket->sendTextMessage(username);
+    _webSocket->sendTextMessage(password);
+}
+
+void PushNotifications::onWebSocketDisconnected() { }
+
+void PushNotifications::onWebSocketTextMessageReceived(const QString &message)
+{
+    if (message == "notify_file") {
+        emit filesChanged(_account);
+    } else if (message == "authenticated") {
+        // TODO: Log
+    } else if (message == "err: Invalid credentials") {
+        connectWebSocket();
+    }
+}
+
+void PushNotifications::onWebSocketError(QAbstractSocket::SocketError error)
+{
+    switch (error) {
+        // TODO: Maybe few more cases go here that need an reconnect ?
+    case QAbstractSocket::NetworkError:
+        connectWebSocket();
+        break;
+    default:;
+        // TODO: Log unhandled errors
+        Q_ASSERT(false && "Unexpected network error not handled");
+    }
+}
+
+void PushNotifications::onWebSocketSslErrors(const QList<QSslError> &errors)
+{
+    // TODO: What to do with them?
+}
+
+void PushNotifications::setupFilesPushNotifications()
+{
+    // Check that push notifications for files are available
+    if (!pushNotificationsAvailable().testFlag(OCC::PushNotificationType::Files))
+        return;
+
+    // Disconnect signal handlers
+    disconnect(_webSocket.data(), &AbstractWebSocket::connected, this, &PushNotifications::onWebSocketConnected);
+    disconnect(_webSocket.data(), &AbstractWebSocket::connected, this, &PushNotifications::onWebSocketConnected);
```

Comment:
@FlexW I guess this line is a duplicate of a preceding?

Reply:
@FlexW Or, maybe you meant to have this one?

disconnect(_webSocket.data(), &AbstractWebSocket::disconnected, this, &PushNotifications::onWebSocketDisconnected);

Reply:
Yes you're right but I have already catched that one:)